### PR TITLE
Made sure not to delete handles of dart objects if the isolate has been deleted

### PR DESCRIPTION
**Pull request by gaaclarke**
This fixes a crash but can result in memory living longer than expected.  The memory fix will require a larger engineering effort.  I think we should land this in the meantime.  See the attached issue for a detailed explanation about what is happening here.

fixes https://github.com/flutter/flutter/issues/79335

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

### CODE REVIEW

File: third_party/tonic/dart_persistent_value.cc [hunk 1]
```diff
@@ -43,12 +43,19 @@ void DartPersistentValue::Clear() {
...
-  if (Dart_CurrentIsolateGroup()) {
-    Dart_DeletePersistentHandle(value_);
-  } else {
-    DartIsolateScope scope(dart_state->isolate());
-    Dart_DeletePersistentHandle(value_);
+  /// TODO(tbd): Remove the handle even if the isolate is shutting down.  This
+  /// may cause memory to stick around until the isolate group is destroyed.
+  /// Currently if DartState::IsShuttingDown == true, this code will crash when
+  /// binding the isolate.
+  if (!dart_state->IsShuttingDown()) {
```
Comment (gaaclarke) [inline third_party/tonic/dart_persistent_value.cc:50]: This matches the logic in DartWeakPersistentValue.

**Comment by gaaclarke** [comment]
@xster @mkustermann what do you think about this approach for now?  I don't know what our policy is for editing tonic is, I'll come up with a unit test for the DartIsolate stuff, too.

**Comment by mkustermann** [comment]
> @xster @mkustermann what do you think about this approach for now? I don't know what our policy is for editing tonic is, I'll 
> come up with a unit test for the DartIsolate stuff, too.

This LGTM. Checking for shutdown condition is solving the exact same issue as in weak persistent handles, so it should have the same solution.

The original commit that added this condition to weak handles had some simple tests in it - maybe you can add a test case there, see ccdb6819ce536899d61115a8c1de89aeb8fc8d41

File: third_party/tonic/dart_persistent_value.cc [hunk 2]
```diff
@@ -43,12 +43,19 @@ void DartPersistentValue::Clear() {
...
   }
 
-  if (Dart_CurrentIsolateGroup()) {
-    Dart_DeletePersistentHandle(value_);
-  } else {
-    DartIsolateScope scope(dart_state->isolate());
-    Dart_DeletePersistentHandle(value_);
+  /// TODO(80155): Remove the handle even if the isolate is shutting down.  This
+  /// may cause memory to stick around until the isolate group is destroyed.
+  /// Currently if DartState::IsShuttingDown == true, this code will crash when
```
Comment (xster) [inline third_party/tonic/dart_persistent_value.cc:48]: From the perspective of someone reading this in the future, this should read "Previously" right?

**Review by xster — APPROVED** [review]

File: third_party/tonic/dart_persistent_value.cc (hunk 2 above)
Comment (gaaclarke) [reply third_party/tonic/dart_persistent_value.cc]: I was referring to the code inside of the branch, that will crash.  Reworded it.

**Comment by sroddy** [comment]
@gaaclarke is it possible to merge this?

**Comment by gaaclarke** [comment]
Yea, the bot was suppose to do it automatically, sorry.

**Comment by sroddy** [comment]
@gaaclarke is it possible to have this cherry-picked to the upcoming stable hot fix as proposed by @mkustermann?

**Comment by xster** [comment]
It already is cherry picked in https://github.com/flutter/engine/pull/25595

**Pull request merged**